account_invoice_accrual/account_invoice.py

```diff
@@ -0,0 +1,323 @@
+            .move_line_get_item(cr, uid, line, context=context)
+        if context.get('move_accrual', False) and context.get('type', False):
+            mapping_obj = self.pool['account.cutoff.mapping']
+            if context.get('type', False) in ['out_invoice', 'out_refund']:
+                mapping_type = 'accrued_revenue'
+            else:
+                mapping_type = 'accrued_expense'
+            mapping = mapping_obj.\
+                _get_mapping_dict(cr, uid, line.invoice_id.company_id.id,
+                                  mapping_type, context=context)
+            res['account_id'] = mapping.get(line.account_id.id,
+                                            line.account_id.id)
+        return res
+
+
+class account_invoice(orm.Model):
+    _inherit = "account.invoice"
+
+    _columns = {
+        'accrual_move_id': fields.many2one(
+            'account.move', 'Accrual Journal Entry',
+            readonly=True, ondelete='restrict', copy=False,
+            help="Link to the Accrual Journal Items."),
+        'accrual_move_name': fields.char('Accrual Journal Entry', size=64,
+                                         readonly=True, copy=False),
+        'to_be_reversed': fields.related(
+            'accrual_move_id', 'to_be_reversed', type='boolean',
+            relation='account.move', string='To be reversed',
+            store=False, readonly=True),
+    }
+
+    def reverse_invoice(self, cr, uid, ids, context=None):
+        # get the list of invoice to reverse
+        if context is None:
+            context = {}
+        ids_to_reverse = []
+        move_ids_to_unlink = []
+        invoice_to_unlink_move = []
+        period_id = False
+        for invoice in self.browse(cr, uid, ids, context=context):
+            if invoice.accrual_move_id:
+                accrual_period_id = invoice.accrual_move_id.period_id.id
+                if invoice.state not in ('draft', 'cancel'):
+                    period_id = invoice.move_id.period_id.id
+                if (not period_id or period_id == accrual_period_id) and \
+                        (not invoice.move_id.id or
+                         invoice.move_id.state == 'draft'):
+                    move_ids_to_unlink.append(invoice.accrual_move_id.id)
+                    invoice_to_unlink_move.append(invoice.id)
+                else:
+                    ids_to_reverse.append(invoice.id)
+        # call reverse method
+        if ids_to_reverse:
+            wiz_obj = self.pool.get("account.move.reverse")
+            wiz_context = dict(context, active_ids=ids_to_reverse,
+                               active_model="account.invoice")
+            wizard_id = wiz_obj.create(cr, uid, {}, context=wiz_context)
+            wiz_obj.action_reverse(cr, uid, [wizard_id], context=wiz_context)
+        if move_ids_to_unlink:
+            self.write(cr, uid, invoice_to_unlink_move,
+                       {'accrual_move_id': False})
```

Comment:
I'm not sure we should reset accrual_move_id when reversing. Because if the user reverses the journal entry without going through the reversal button on the invoice, it will not be reset. So that would give a slightly different behaviour, potentially confusing.